src/include/optimizer/statistics/selectivity_util.h

```diff
@@ -0,0 +1,160 @@
+ * used by this class to compute selectivity.
+ */
+class SelectivityUtil {
+ public:
+  /**
+   * Compute selectivity of a condition
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double ComputeSelectivity(common::ManagedPointer<NewColumnStats<T>> column_stats,
+                                   const ValueCondition &condition);
+
+  /**
+   * Computes selectivity of the Less Than condition
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double LessThan(common::ManagedPointer<NewColumnStats<T>> column_stats, const ValueCondition &condition) {
+    double res = LessThanOrEqualTo(column_stats, condition) - Equal(column_stats, condition);
+    return std::max(std::min(res, 1.0), 0.0);
+  }
+
+  /**
+   * Computes Less Than Or Equal To Selectivity
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double LessThanOrEqualTo(common::ManagedPointer<NewColumnStats<T>> column_stats,
+                                  const ValueCondition &condition);
+
+  /**
+   * Computes Greater Than Selectivity
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double GreaterThan(common::ManagedPointer<NewColumnStats<T>> column_stats, const ValueCondition &condition) {
+    double res = 1 - LessThanOrEqualTo(column_stats, condition) - column_stats->GetFracNull();
+    return std::max(std::min(res, 1.0), 0.0);
+  }
+
+  /**
+   * Computes Greater Than Or Equal To Selectivity
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double GreaterThanOrEqualTo(common::ManagedPointer<NewColumnStats<T>> column_stats,
+                                     const ValueCondition &condition) {
+    double res = 1 - LessThan(column_stats, condition) - column_stats->GetFracNull();
+    return std::max(std::min(res, 1.0), 0.0);
+  }
+
+  /**
+   * Computes Equal Selectivity
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double Equal(common::ManagedPointer<NewColumnStats<T>> column_stats, const ValueCondition &condition);
+
+  /**
+   * Computes Not Equal Selectivity
+   * @param column_stats Column Statistics
+   * @param condition ValueCondition
+   * @returns selectivity
+   */
+  template <typename T>
+  static double NotEqual(common::ManagedPointer<NewColumnStats<T>> column_stats, const ValueCondition &condition) {
+    double res = 1 - Equal(column_stats, condition) - column_stats->GetFracNull();
+    return std::max(std::min(res, 1.0), 0.0);
```

Comment:
Since I am dealing with double values, there are places where the selectivity goes a little below 0 (super small amount) so I think it is better to just clip it. If the clipping is done, there is no need for the assert. In places where I can be sure, there are already ASSERT statements.

Reply:
Can we do an assert for values like (-0.1, 1.1). I'm worried that overflow or underflow errors might go unnoticed ?